lib/apiservers/engine/backends/layer_cache.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vicbackends
+
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/vmware/vic/lib/apiservers/portlayer/client/storage"
+	"github.com/vmware/vic/lib/apiservers/portlayer/models"
+	"github.com/vmware/vic/lib/guest"
+)
+
+// LayerCache is an in-memory cache of image layers. It is refreshed at startup
+// by a call to the portlayer. It is updated when new layers are pulled or
+// layers are deleted.
+type LayerCache struct {
+	sync.Mutex
+
+	// cache maps layer ID to layer data
+	cache map[string]*models.Image
+}
+
+func NewLayerCache() *LayerCache {
+	return &LayerCache{
+		cache: make(map[string]*models.Image),
+	}
+}
+
+// Update adds new layers to the cache
+func (c *LayerCache) Update() error {
+	c.Lock()
+	defer c.Unlock()
+
+	log.Debugf("Updating layer cache...")
+
+	client := PortLayerClient()
+	if client == nil {
+		return fmt.Errorf("vicbackends.hydrateImageCache failed to create a portlayer client")
+	}
+
+	host, err := guest.UUID()
+	if host == "" {
+		host, err = os.Hostname()
+	}
+	if err != nil {
+		return fmt.Errorf("Unexpected error getting hostname: %s", err)
+	}
+
+	// attempt to create the image store if it doesn't exist
+	store := &models.ImageStore{Name: host}
+	_, err = client.Storage.CreateImageStore(
+		storage.NewCreateImageStoreParams().WithBody(store),
+	)
+
+	if err != nil {
+		if _, ok := err.(*storage.CreateImageStoreConflict); ok {
```

Comment:
I probably should change this to just overwrite existing layers in the cache to cover the case of an image having been updated.